### DISCUSSION TITLE
Best way to integrate password store

### DISCUSSION THREAD

**Discussion by btrepp** [argoproj/argo-cd#9061] [Q&A] [answered]
I'm trying to figure out a good secrets story for my home lab. Currently I use passwordstore and ansible, which works really well for the 'base' components.

It's a bit more complicated with argocd though. I like the ability of kustomize to hash the name, and have a proof of concept with a exec plugin.

It gets a bit confusing around having kustomize plugins in the argocd repo server though. It would need to have gpg, a key, and my plugin installed, it would also need to git pull a different 'secrets' repo.

Is there any reasons not to extend the repo server this way?. 

When the repo server syncs. Is it based on the manifest output, or the git hash?. If it's on the manifests I would gain the benefit that password changes would result in new deployments. Which is nice.

**Comment by btrepp** [comment]
It might be nice if I can somehow tie the kustomize to the pass git hash. That would be even more determinism. Eg it's passwords at this point in time, and probably help argocd resolve.

The trick would be seeing if I can make kustomize depend on resources outside of kustomize manifests.

**Reply by btrepp** [reply]
Current trick I think will be figuring out the repo server and security. If I mount the gpg keys. It means anything that can potentially access the gpg keys. Eg a mailicious kustomize exec script.

**Reply by crenshaw-dev** [reply]
I think holding secrets securely on the repo-server is still a not-entirely-solved problem (https://github.com/argoproj/argo-cd/issues/9083). One approach is to mount the secret and then make sure you really really trust your plugin implementation to keep it secure. Another is to use something like SealedSecrets to push the problem only the cluster and out of Argo CD.

**Reply by btrepp** [reply]
At the moment I'm thinking of writing a CMP plugin to handle my case. 

At a simple level if the 'admin' controls the application, then the variables on the application will help limit what a repo could access. Though it would ban the app of apps pattern for any 'less trusted' tenants.

For my use case this would be acceptable until #9083 has a bit of a stronger path

**Reply by crenshaw-dev** [reply]
Sweet, would be interested to see the CMP if you're able to open-source it. 

App-of-apps is difficult to lock down. I think the safest way to do it in a multi-tenant environment is by giving each tenant their own namespaced Argo CD instance and destination cluster.

**Comment by todaywasawesome** [comment]
For a homelab I would think that sealed secrets would be a great solution. @btrepp 
See this article from @kostis-codefresh  https://codefresh.io/about-gitops/handle-secrets-like-pro-using-gitops/

**Reply by btrepp** [reply]
I had run sealed secrets before. It is not bad. Though can be a bit clunky.

At the moment trying to unify on the one secret store (eg password store) which also helps create the initial cluster.

**Comment by nastacio** [comment]
I would look into external secrets: https://external-secrets.io/v0.5.1/provider-kubernetes/

**Reply by btrepp** [reply]
Thanks :). I did have a go with some of the externap secrets. Seems you need to have another tool holding the secrets. I did get a vault running, but it got pretty complex.

**Comment by btrepp** [comment, marked as answer]
I think I was the first person to want to integrate pass and argocd :)

So there was no best practice. There still probably isn't, but at least I've written a tool and plugin for anyone who might want to do this.

https://git.sr.ht/~btrepp/argo-pass-sidecar
